scheduler/cloudformation/cloudformation.go

```diff
@@ -268,11 +271,129 @@ func (s *Scheduler) submit(ctx context.Context, tx *sql.Tx, app *scheduler.App,
 		if err := <-done; err != nil {
 			return err
 		}
+		done = make(chan error, 1)
+		s.waitForDeploymentToStabilize(ctx, stackName, ss, done)
+		if err = <-done; err != nil {
+			return err
+		}
 	}
 
 	return nil
 }
 
+func (s *Scheduler) waitForDeploymentToStabilize(ctx context.Context, name string, ss scheduler.StatusStream, done chan error) {
```

Comment:
We might be able to simplify this method by returning a channel where we send stabilization events on. Something like this:

```go
// waitForDeploymentsToStabilize returns a channel that you can range over to
// wait for multiple ECS deployments to stabilize. The channel is closed
// when one of the following occurs:
//
//	1. All deployments enter a "stable" state.
//	2. Any deployment enters an inactive state.
func (s *Scheduler) waitForDeploymentsToStabilize(deployments []ecsDeployment) <-chan deploymentStatus {
	ch := make(chan deploymentStatus)

	go func(deployments []ecsDeployment) {
		for len(deployments) > 0 {
			// Describe the services

			for i, d := range deployments {
				if stable {
					ch <- deploymentStatus{d, "stable"}
					// Remove the deployment, since it's
					// stable. No need to keep checking it.
					deployments = append(deployments[:i], deployments[i+1]...)
				} else if primary {
					// do nothing
				} else {
					ch <- deploymentStatus{d, "inactive"}
					close(ch)
					return
				}
			}
		}

		// All done
		close(ch)
	}(deployments[:])

	return ch
}
```

Then we can generate the list of deployments, and publish to the status stream as deployments change state.

```go
deploymentStatuses := s.waitForDeploymentsToStabilize(deployments)
for status := range deploymentStatuses {
	scheduler.Publish("Service %s became %s", status.service, status)
}

// Done
```